url1cbaffb197_foxking0416_AlgorithmTraining_tgz-pJ8-algorithm_leetcode_MainLeetJ8/src/datastructure/avltree/AVLTree.java

```diff
@@ -0,0 +1,223 @@
+				currentNode.setRightLeafNode(newNode);
+			}
+			else {
+				insertNode(currentNode.getRightLeafNode(), newNode);
+			}
+		}
+		
+		checkBalance(currentNode);
+	}
+	
+	
+	@Override
+	public void remove(TreeNode removeNode){
+		TreeNode nodeToRemoveNode =  this.rootNode;
+		TreeNode parentNode = null;
+		Stack<TreeNode> pathStack = new Stack<TreeNode>();
+		pathStack.add(this.rootNode);
+		
+		while(nodeToRemoveNode != null && nodeToRemoveNode != removeNode){
+			parentNode = nodeToRemoveNode;
+			if(removeNode.val < nodeToRemoveNode.val){
+				nodeToRemoveNode = parentNode.getLeftLeafNode();
+			}
+			else{
+				nodeToRemoveNode = parentNode.getRightLeafNode();
+			}
+			pathStack.push(nodeToRemoveNode);
+		}
+		
+		
+		if(nodeToRemoveNode == null)
+			return;
+		
+		//parentNode = findParentNode(removeNode, this.rootNode);
+		
+		
+		if(count == 1)
+			this.rootNode = null;//remove the only one element of this tree
+		else{
+			if(nodeToRemoveNode.getLeftLeafNode() == null && nodeToRemoveNode.getRightLeafNode() == null){//remove the leaf node
+				if(nodeToRemoveNode.val < parentNode.val){
+					parentNode.setLeftLeafNode(null);
+				}
+				else {
+					parentNode.setRightLeafNode(null);
+				}
+			}
+			else if (nodeToRemoveNode.getLeftLeafNode() != null && nodeToRemoveNode.getRightLeafNode() == null){//remove the node with only left leaf node
+				if(nodeToRemoveNode.val < parentNode.val){
+					parentNode.setLeftLeafNode(nodeToRemoveNode.getLeftLeafNode());
+				}
+				else{
+					parentNode.setRightLeafNode(nodeToRemoveNode.getLeftLeafNode());
+				}
+			}
+			else if (nodeToRemoveNode.getLeftLeafNode() == null && nodeToRemoveNode.getRightLeafNode() != null){//remove the node with only right leaf node
+				if(nodeToRemoveNode.val < parentNode.val){
+					parentNode.setLeftLeafNode(nodeToRemoveNode.getRightLeafNode());
+				}
+				else{
+					parentNode.setRightLeafNode(nodeToRemoveNode.getRightLeafNode());
+				}	
+			}
+			else{//remove the node with two leaf nodes
+				TreeNode largestNodeFromLeft = nodeToRemoveNode.getLeftLeafNode();
+				while(largestNodeFromLeft.getRightLeafNode() != null){
+					largestNodeFromLeft = largestNodeFromLeft.getRightLeafNode();
+				}
+				
+				
+				TreeNode tempParent = findParentNode(largestNodeFromLeft, this.rootNode);
+
+				
+				largestNodeFromLeft.setRightLeafNode(nodeToRemoveNode.getRightLeafNode());
+				if(largestNodeFromLeft.val < tempParent.val){
+					largestNodeFromLeft.setLeftLeafNode(nodeToRemoveNode.getLeftLeafNode().getLeftLeafNode());	
+				}
+				else{
+					largestNodeFromLeft.setLeftLeafNode(nodeToRemoveNode.getLeftLeafNode());	
+					tempParent.setRightLeafNode(null);
+				}
+				
+				if(parentNode != null){
+					if(nodeToRemoveNode.val < parentNode.val){
+						parentNode.setLeftLeafNode(largestNodeFromLeft);
+					}
+					else{
+						parentNode.setRightLeafNode(largestNodeFromLeft);
+					}
+				}
+				else{
+					this.rootNode = largestNodeFromLeft;
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 104:114, 161:171, 187:197.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.